File: Sobel_Operator.m
% Initialization
close all;
clear all;
clc;

% Read and display the input image
image1 = imread('grocery.jpg');
imshow(image1);
title('Original Image');

% Convert the image to grayscale
Img = rgb2gray(image1); % converting the rgb image to a gray level image
Img = im2double(Img);

% Define the Sobel filters for x and y directions
sobel_x = [-1 0 1; -2 0 2; -1 0 1]; 
sobel_y = [1 2 1; 0 0 0; -1 -2 -1];

% Apply the Sobel filters to get gradients
grad_x = conv2(Img, sobel_x, 'same');
grad_y = conv2(Img, sobel_y, 'same');

% Compute the gradient magnitude
grad_magnitude = sqrt(grad_x.^2 + grad_y.^2);

% Display the gradients and the gradient magnitude
figure;
subplot(1, 3, 1);
imshow(grad_x, []);
title('Gradient in X direction');

subplot(1, 3, 2);
imshow(grad_y, []);
title('Gradient in Y direction');

subplot(1, 3, 3);
imshow(grad_magnitude, []);
title('Gradient Magnitude');

File: Sobel_Operator.py
import cv2 as cv
import numpy as np
import matplotlib.pyplot as plt

img = cv.imread('sports_.jpg')
cv.imshow('Original', img)

# converting the image to gray scale
gray_img = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
cv.imshow('Grayscale', gray_img)

# Laplacian Method
laplacian = cv.Laplacian(gray_img, cv.CV_64F)
laplacian = np.uint8(np.abs(laplacian))# takes in source image (gray scale), data depth
cv.imshow('Laplacian Image', laplacian)

# Sobel Method
sobel_x = cv.Sobel(gray_img, cv.CV_64F, 1, 0)
sobel_y = cv.Sobel(gray_img, cv.CV_64F, 0, 1)
cv.imshow('Sobel_x', sobel_x)
cv.imshow('Sobel_y', sobel_y)

combined_sobel = cv.bitwise_or(sobel_x, sobel_y)
cv.imshow('Combined sobel', combined_sobel)

# including canny for comparison
canny_img = cv.Canny(gray_img, 150, 175)
cv.imshow('Canny', canny_img)


plt.figure(figsize=(20,10))
plt.subplot(1,3,1)
plt.title('Laplacian Image')
plt.imshow(laplacian, cmap='gray')

plt.subplot(1,3,2)
plt.title('Sobel Image')
plt.imshow(combined_sobel, cmap='gray')

plt.subplot(1,3,3)
plt.title('Canny Image')
plt.imshow(canny_img, cmap='gray')

plt.tight_layout()
plt.show()


cv.waitKey(0)